guides/api-guideline.md

```diff
@@ -0,0 +1,135 @@
+|-----------------|---------------------------------------------------------------|
+| `string`        | A sequence of characters                                      |
+| `intX`          | A signed integer of X bits                                    |
+| `uintX`         | An unsigned integer of X bits                                 |
+| `int`           | A signed integer of 32 bits                                   |
+| `long`          | A signed integer of 64 bits                                   |
+| `double`        | A native floating-point number in 64-bit base-2 format        |
+| `decimal`       | A decimal number with arbitrary precision                     |
+| `bool`          | A boolean value                                               |
+| `bytes`         | A sequence of bytes                                           |
+| `list<TYPE>`    | A list of elements of type TYPE                               |
+| `set<TYPE>`     | A set of elements of type TYPE                                |
+| `map<TYPE>`     | An immutable map of elements of type TYPE                     |
+| `date`          | A date value (ISO 8601 calendar date)                         |
+| `time`          | A time value without date or timezone (nanosecond precision)  |
+| `dateTime`      | A date and time value without timezone (nanosecond precision) |
+| `zonedDateTime` | A date and time value with timezone (nanosecond precision)    |
+
+### Enumerations
+
+Enumerations can be defined using the following syntax:
+```
+enum EnumName {
+    VALUE1
+    VALUE2
+}
+```
+
+### Attribute annotations
+
+Attribute annotations can be used to provide additional information about attributes in complex data types.
+The following annotations should be used:
+- `@immutable`: Indicates that the field is immutable and cannot be changed after creation.
+- `@optional`: Indicates that the field is optional and can be null or undefined.
+- `@default(value)`: Indicates that the field has a default value.
+- `@deprecated`: Indicates that the field is deprecated and should not be used.
+- `@removed`: Indicates that the field must be removed.
+- `@min(value)`: Indicates the minimum value for numeric fields.
+- `@max(value)`: Indicates the maximum value for numeric fields.
+- `@minLength(value)`: Indicates the minimum length for string fields.
+- `@maxLength(value)`: Indicates the maximum length for string fields.
+- `@pattern(regex)`: Indicates a regex pattern that the string field must match.
+
+### Method annotations
+
+Method annotations can be used to provide additional information about methods.
+The following annotations should be used:
+- `@async`: Indicates that the method is asynchronous and returns a promise or future.
+- `@throws(error-type)`: Indicates that the method can throw an exception/error.
+- `@deprecated`: Indicates that the method is deprecated and should not be used.
+- `@removed`: Indicates that the method must be removed.
```

Comment:
I'm trying to think of a time when `@deprecated` or `@removed` would ever be used in our design docs (same for attributes above too). If we are deprecating/removing methods, I think it would make sense to explicitly call them out under a `Deprecate/Remove` heading or something